tests/subsys/net/lib/nrf_cloud/cloud/src/main.c

```diff
@@ -0,0 +1,2680 @@
+		"nrf_cloud lib should be in the connected state after successfully connected");
+
+	int ret = nrf_cloud_uninit();
+
+	/* Wait for the disconnected event */
+	while (atomic_get(&cloud_connected_t)) {
+		k_sleep(K_MSEC(200));
+	}
+
+	zassert_ok(ret, "return should be 0 when uninit success");
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state after uninit");
+}
+
+/* Verify that nrf_cloud_uninit function succeeds
+ * after initialized without cloud connection.
+ */
+ZTEST(nrf_cloud_test, test_uninit_success_after_init_no_connection)
+{
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	struct nrf_cloud_init_param init_param_t = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	/* Init the cloud successfully */
+	nct_init_fake.custom_fake = fake_nct_init__succeeds;
+	nfsm_init_fake.custom_fake = fake_nfsm_init__succeeds;
+	nrf_cloud_codec_init_fake.custom_fake = fake_nrf_cloud_codec_init__succeeds;
+	(void)nrf_cloud_init(&init_param_t);
+
+	zassert_equal(STATE_INITIALIZED, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the initialized state after initialization");
+
+	/* Uninit successfully */
+	int ret = nrf_cloud_uninit();
+
+	zassert_ok(ret, "return should be 0 when uninit success");
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state after uninit");
+}
+
+/* Verify that nrf_cloud_uninit function will fail when the
+ * poll thread does not become inactive after 10 seconds.
+ * Ignored if CONNECTION_POLL_THREAD is disabled.
+ */
+ZTEST(nrf_cloud_test, test_uninit_poll_thread_inactive_timer_expired)
+{
+	Z_TEST_SKIP_IFNDEF(CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD);
+	zassert_equal(STATE_IDLE, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the idle state (uninitialized) at the start of test");
+
+	struct nrf_cloud_init_param init_param_t = {
+		.event_handler = event_handler,
+		.client_id = NULL,
+	};
+
+	/* Init the cloud successfully */
+	nct_init_fake.custom_fake = fake_nct_init__succeeds;
+	nfsm_init_fake.custom_fake = fake_nfsm_init__succeeds;
+	nrf_cloud_codec_init_fake.custom_fake = fake_nrf_cloud_codec_init__succeeds;
+	(void)nrf_cloud_init(&init_param_t);
+
+	zassert_equal(STATE_INITIALIZED, nfsm_get_current_state(),
+		"nrf_cloud lib should be in the initialized state after initialization");
+
+	/* Connect to the cloud successfully */
+	nct_connect_fake.custom_fake = fake_nct_connect__succeeds;
+	nct_disconnect_fake.custom_fake = fake_nct_disconnect__succeeds;
+	nct_socket_get_fake.custom_fake = fake_nct_socket_get_data;
+	nct_keepalive_time_left_fake.custom_fake = fake_nct_keepalive_time_left_long_timeout;
+	nct_process_fake.custom_fake = fake_nct_process__succeeds;
+	(void)nrf_cloud_connect();
+
+	/* Wait for the connected event */
+	while (!atomic_get(&cloud_connected_t)) {
+		k_sleep(K_MSEC(200));
```

Comment:
Wouldn't it be better to timeout here eventually, rather than potentially hang forever?

Reply:
A similar loop appears in many tests below.

Reply:
Haven't had a chance to do full review yet, sorry, but in addition to implementing a timeout, consider using a semaphore instead of atomic type, that way you can just await the semaphore instead of using a while loop at all

Reply:
I will think of another method for this

Reply:
Added a counter for timeout, but that also needs a while loop, so I kept the atomic type

Reply:
I mean you could avoid the while-loop altogether if you use a semaphore

Reply:
yeah, you're right. There IS a timeout when taking a semaphore.